### PR TITLE
[Merged by Bors] - get proper texture format after the renderer is initialized, fix #3897

### LINKED ISSUE #3897: OS-independent texture format setting in swapchain
## What problem does this solve or what need does it fill?

The current `bevy_render` crate assumes the supported texture format based on OS. This may lead to issues, e.g. Nvidia on Wayland not supporting the expected `Bgra8UnormSrgb` format (only supporting `Bgra8Unorm`).

## What solution would you like?

Perform platform-independent querying of supported texture formats in swapchain, with a preferred one then being picked for `Surface:: configure`.

## What alternative(s) have you considered?

(In my case) using Xwayland by disabling the `wayland` feature works, but causes blurriness with fractional scaling.

## Additional context

My current (incomplete) fix can be found at https://github.com/a-b-c-1-2-3/bevy on branch `support-non-srgb`, but this only allows for a feature to disable sRGB, not proper runtime querying of supported formats.

### PR DESCRIPTION
**Pull request by VitalyAnkh**
# Objective
There is no Srgb support on some GPU and display protocols with `winit` (for example, Nvidia's GPUs with Wayland). Thus `TextureFormat::bevy_default()` which returns `Rgba8UnormSrgb` or `Bgra8UnormSrgb` will cause panics on such platforms. This patch will resolve this problem. Fix https://github.com/bevyengine/bevy/issues/3897.

## Solution

Make `initialize_renderer` expose `wgpu::Adapter` and `first_available_texture_format`, use the `first_available_texture_format` by default.

## Changelog

* Fixed https://github.com/bevyengine/bevy/issues/3897.

### CODE REVIEW

File: crates/bevy_render/src/lib.rs [hunk 1]
```diff
@@ -148,14 +148,16 @@ impl Plugin for RenderPlugin {
                 compatible_surface: surface.as_ref(),
                 ..Default::default()
             };
-            let (device, queue, adapter_info) = futures_lite::future::block_on(
+            let (device, queue, adapter_info, render_adapter) = futures_lite::future::block_on(
                 renderer::initialize_renderer(&instance, &options, &request_adapter_options),
             );
             debug!("Configured wgpu adapter Limits: {:#?}", device.limits());
             debug!("Configured wgpu adapter Features: {:#?}", device.features());
+            debug!("Now the RenderPlugin runs VitalyR");
```
Comment (superdump) [inline crates/bevy_render/src/lib.rs:156]:
```suggestion
```

File: crates/bevy_render/src/view/window.rs [hunk 2]
```diff
@@ -161,7 +160,7 @@ pub fn prepare_windows(
             });
 
         let swap_chain_descriptor = wgpu::SurfaceConfiguration {
-            format: TextureFormat::bevy_default(),
+            format: surface.get_supported_formats(&render_adapter)[0],
```
Comment (superdump) [inline crates/bevy_render/src/view/window.rs:163]: This will panic if the returned `Vec` is empty. It would probably be better to still panic, but to panic with a more useful error message.

**Review by superdump — CHANGES_REQUESTED** [review]
We need to think about the consequences of adding and exposing the `Adapter`.

This is a good change that needs to happen.

I think further than this, probably bevy_default should be removed and replaced when the tonemapping PR lands. @cart

**Comment by VitalyAnkh** [comment]
@superdump Thanks for your kind reply! Now I have removed the unused `dbg!`. The latest code now add `AvailableTextureFormat` resource to the `World` and use it for future use. I know the available `TextureFormat` should be computed per `Surface`, but it's still better than hard-coded values. And then we don't need to expose `Adapter` now but the code still does so now (if Bevy developers think it's useful, could I do it in another PR?). And I add a `GpuImage` field to `UiPipeline` to specify correct `TextureFormat` in `impl<S: SpecializedRenderPipeline> SpecializedRenderPipelines<S>`. The code still needs much tuning, but it could fix #3897 correctly now.

**Comment by VitalyAnkh** [comment]
Oops... It fails on x11 now. I will push more fixes.

**Comment by VitalyAnkh** [comment]
I think this PR could pass all checks now and with this, there won't be any "no available texture formats" error anymore. But the design still needs discussion. For example, should I expose `Adapter` in the `World`? And to make `MeshPipe` `UiPipeline` and `SpritePipeline` get the right `TextureFormat`, I add a `GpuImage` field to them, should I do this? Are there better ways to solve this problem?

File: crates/bevy_render/src/view/window.rs (hunk 2 above)
Comment (VitalyAnkh) [reply crates/bevy_render/src/view/window.rs:163]: Done! `info!` the `AdapterInfo` then `panic!("No supported texture formats found!")`.

File: crates/bevy_render/src/lib.rs (hunk 1 above)
Comment (VitalyAnkh) [reply crates/bevy_render/src/lib.rs]: Deleted.

**Comment by bemyak** [comment]
I've just run into the issue that this PR addresses. Any plans on getting it merged?

**Comment by VitalyAnkh** [comment]
@superdump @alice-i-cecile hi, what do you think about this PR? What should I do to get this merged?

File: crates/bevy_render/src/renderer/mod.rs [hunk 3]
```diff
@@ -97,13 +101,26 @@ pub struct RenderInstance(pub Instance);
 #[derive(Resource, Clone, Deref, DerefMut)]
 pub struct RenderAdapterInfo(pub AdapterInfo);
 
+/// The `AdapterInfo` of the adapter in use by the renderer.
+#[derive(Resource, Clone, Deref, DerefMut)]
+pub struct RenderTextureFormat(pub wgpu::TextureFormat);
+
+#[derive(Resource, Clone, Deref, DerefMut)]
+pub struct AvailableTextureFormats(pub Arc<Vec<wgpu::TextureFormat>>);
```
Comment (alice-i-cecile) [inline crates/bevy_render/src/renderer/mod.rs:109]: Basic doc string please.

File: crates/bevy_render/src/renderer/mod.rs [hunk 4]
```diff
@@ -97,13 +101,26 @@ pub struct RenderInstance(pub Instance);
 #[derive(Resource, Clone, Deref, DerefMut)]
 pub struct RenderAdapterInfo(pub AdapterInfo);
 
+/// The `AdapterInfo` of the adapter in use by the renderer.
```
Comment (alice-i-cecile) [inline crates/bevy_render/src/renderer/mod.rs:104]: This doc string seems wrong.

File: crates/bevy_render/src/renderer/mod.rs [hunk 5]
```diff
@@ -97,13 +101,26 @@ pub struct RenderInstance(pub Instance);
...
+pub struct AvailableTextureFormats(pub Arc<Vec<wgpu::TextureFormat>>);
+
 /// Initializes the renderer by retrieving and preparing the GPU instance, device and queue
 /// for the specified backend.
 pub async fn initialize_renderer(
     instance: &Instance,
     options: &WgpuSettings,
     request_adapter_options: &RequestAdapterOptions<'_>,
-) -> (RenderDevice, RenderQueue, RenderAdapterInfo) {
+) -> (
```
Comment (alice-i-cecile) [inline crates/bevy_render/src/renderer/mod.rs:117]: In my experience returning complex tuple types like this is a footgun and maintenance hazard. Can you refactor this to use a named struct?

File: crates/bevy_sprite/src/render/mod.rs [hunk 6]
```diff
@@ -79,10 +87,57 @@ impl FromWorld for SpritePipeline {
             ],
             label: Some("sprite_material_layout"),
         });
+        let dummy_white_gpu_image = {
```
Comment (alice-i-cecile) [inline crates/bevy_sprite/src/render/mod.rs:90]: The code duplication of this section strikes me as a bit suspicious. Is there a good way to refactor this to avoid needing to remember to add this to each pipeline?

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
First round of feedback left :) This is definitely a real issue, and the fix makes sense.

I'm not familiar enough with the details of the rendering machinery here to suggest exactly how you might fix that code duplication issue, but the current strategy makes me suspect that any custom implementations will run into the same problem.

**Comment by superdump** [comment]
I’ll give this another look over soon. I would like to have reviewed this and other things already but life is what it is. I appreciate your patience.

File: crates/bevy_render/src/renderer/mod.rs (hunk 5 above)
Comment (VitalyAnkh) [reply crates/bevy_render/src/renderer/mod.rs:117]: How about this:
```rust
// at bevy/crates/bevy_render/src/renderer/mod.rs
struct Renderer{
    render_device: RenderDevice,
    render_queue: RenderQueue,
    render_adapter_info: RenderAdapterInfo,
    render_adapter: RenderAdapter,
    available_texture_formats: AvailableTextureFormats,
}
```
Then make `initialize_renderer` return this struct and change all the related code?

File: crates/bevy_render/src/renderer/mod.rs (hunk 4 above)
Comment (VitalyAnkh) [reply crates/bevy_render/src/renderer/mod.rs:104]: fixed

File: crates/bevy_render/src/renderer/mod.rs (hunk 3 above)
Comment (VitalyAnkh) [reply crates/bevy_render/src/renderer/mod.rs:109]: Added.

File: crates/bevy_render/src/renderer/mod.rs (hunk 5 above)
Comment (alice-i-cecile) [reply crates/bevy_render/src/renderer/mod.rs:120]: Yep, that's exactly the sort of thing I had in mind!

**Comment by bemyak** [comment]
Friendly bumping @superdump and @alice-i-cecile to review this PR :heart:

File: crates/bevy_render/src/renderer/mod.rs (hunk 5 above)
Comment (VitalyAnkh) [reply crates/bevy_render/src/renderer/mod.rs:120]: Hi!  Maybe wrapping the return types into a struct over engineering? This function returns a tuple whose every element will be inserted as resource in the world. We only care about every element of the tuple. And before this change,  this function returns a three element tuple. I think theere is no need to change this. How do you think?

Comment (alice-i-cecile) [reply crates/bevy_render/src/renderer/mod.rs:120]: Yeah I'm fine to leave it.

File: crates/bevy_pbr/src/render/mesh.rs [hunk 7]
```diff
@@ -269,8 +267,10 @@ impl FromWorld for MeshPipeline {
             Res<RenderDevice>,
             Res<DefaultImageSampler>,
             Res<RenderQueue>,
+            Res<RenderTextureFormat>,
         )> = SystemState::new(world);
-        let (render_device, default_sampler, render_queue) = system_state.get_mut(world);
+        let (render_device, default_sampler, render_queue, firstly_available_texture_format) =
```
Comment (alice-i-cecile) [inline crates/bevy_pbr/src/render/mesh.rs:272]: Nit: I think gramattically this should be `first_available_texture_format)`

File: crates/bevy_render/src/renderer/mod.rs [hunk 8]
```diff
@@ -88,6 +88,10 @@ pub fn render_system(world: &mut World) {
 #[derive(Resource, Clone, Deref, DerefMut)]
 pub struct RenderQueue(pub Arc<Queue>);
 
+/// This is the adapter that the GPU is using to render.
```
Comment (alice-i-cecile) [inline crates/bevy_render/src/renderer/mod.rs:91]: This doc comment isn't helpful to someone who doesn't know what an adapter is.

We should consider linking to the `wgpu` docs here.

File: crates/bevy_render/src/renderer/mod.rs [hunk 9]
```diff
@@ -97,13 +101,29 @@ pub struct RenderInstance(pub Instance);
 #[derive(Resource, Clone, Deref, DerefMut)]
 pub struct RenderAdapterInfo(pub AdapterInfo);
 
+/// The texture format used by the renderer.
```
Comment (alice-i-cecile) [inline crates/bevy_render/src/renderer/mod.rs:104]: Same as above; this doc comment is fairly circular. Try to briefly explain what a texture format is :)

File: crates/bevy_render/src/renderer/mod.rs [hunk 10]
```diff
@@ -97,13 +101,29 @@ pub struct RenderInstance(pub Instance);
...
+/// Initially it's the first texture format in `AvailableTextureFormats`
+#[derive(Resource, Clone, Deref, DerefMut)]
+pub struct RenderTextureFormat(pub wgpu::TextureFormat);
+
+/// Available texture formats on a given platform. Will be inserted
```
Comment (alice-i-cecile) [inline crates/bevy_render/src/renderer/mod.rs:109]: This needs to link back and forth with the struct above. It's easy to figure out their relation in the source, but many users will be interacting with this through pure docs.

File: crates/bevy_render/src/view/window.rs (hunk 2 above)
Comment (alice-i-cecile) [inline crates/bevy_render/src/view/window.rs:163]: This line is hard to follow in context. IMO we should have a `surface.first_detected_texture` or something that gets called here and has solid docs.

File: crates/bevy_render/src/renderer/mod.rs (hunk 9 above)
Comment (devil-ira) [inline crates/bevy_render/src/renderer/mod.rs:104]:
```suggestion
/// The [`TextureFormat`](wgpu::TextureFormat) used for render targets.
```

File: crates/bevy_render/src/renderer/mod.rs (hunk 8 above)
Comment (devil-ira) [inline crates/bevy_render/src/renderer/mod.rs:91]:
```suggestion
/// The handle to the physical device being used for rendering.
/// See [`wgpu::Adapter`] for more info.
```

File: crates/bevy_render/src/view/window.rs (hunk 2 above)
Comment (devil-ira) [reply crates/bevy_render/src/view/window.rs:163]: IMO the docs on [get_supported_formats](https://docs.rs/wgpu/latest/wgpu/struct.Surface.html#method.get_supported_formats) are good enough.
Handling the case of `get_supported_formats` returning an empty Vec more gracefully would be nice.

File: crates/bevy_render/src/renderer/mod.rs [hunk 11]
```diff
@@ -97,13 +101,29 @@ pub struct RenderInstance(pub Instance);
...
+/// as `Resource` after the renderer is initialized.
```
Comment (devil-ira) [inline crates/bevy_render/src/renderer/mod.rs:110]:
```suggestion
/// The available [`TextureFormat`](wgpu::TextureFormat)s on the [`RenderAdapter`].
/// Will be inserted as a `Resource` after the renderer is initialized.
```

File: crates/bevy_render/src/renderer/mod.rs (hunk 8 above)
Comment (VitalyAnkh) [reply crates/bevy_render/src/renderer/mod.rs:91]: Thanks!

File: crates/bevy_render/src/renderer/mod.rs (hunk 9 above)
Comment (VitalyAnkh) [reply crates/bevy_render/src/renderer/mod.rs]: Thanks!

File: crates/bevy_render/src/view/window.rs (hunk 2 above)
Comment (VitalyAnkh) [reply crates/bevy_render/src/view/window.rs]: > Handling the case of get_supported_formats returning an empty Vec more gracefully would be nice.

Done.

File: crates/bevy_render/src/renderer/mod.rs (hunk 11 above)
Comment (VitalyAnkh) [reply crates/bevy_render/src/renderer/mod.rs]: Thanks!

File: crates/bevy_render/src/renderer/mod.rs (hunk 10 above)
Comment (VitalyAnkh) [reply crates/bevy_render/src/renderer/mod.rs]: Done.

**Comment by VitalyAnkh** [comment]
@alice-i-cecile @devil-ira Hi! Following your kind reviews, I have updated this PR. All the problems you proposed should be resolved. Please let me know if there is anything that should be improved, thanks!

File: crates/bevy_pbr/src/render/mesh.rs (hunk 7 above)
Comment (VitalyAnkh) [reply crates/bevy_pbr/src/render/mesh.rs]: Done. :)

File: crates/bevy_render/src/renderer/mod.rs (hunk 8 above)
Comment (VitalyAnkh) [reply crates/bevy_render/src/renderer/mod.rs]: Done by following @devil-ira's suggestion.

File: crates/bevy_render/src/renderer/mod.rs (hunk 9 above)
Comment (VitalyAnkh) [reply crates/bevy_render/src/renderer/mod.rs]: Done.

**Review by devil-ira — APPROVED** [review]

**Review by alice-i-cecile — APPROVED** [review]

**Comment by VitalyAnkh** [comment]
The latest force push makes `prepare_view_targets` use `Res<RenderTextureFormat>` other than `Res<AvailableTextureFormats>` to get proper texture format. It doesn't touch codes that have been reviewed.

**Comment by devil-ira** [comment]
I had some trouble with refresh rates with a multi-monitor setup, so I switched to wayland and then bumped into this issue myself.
I tried this PR, and while it does run, the colors aren't correct.
poroporo

**Comment by VitalyAnkh** [comment]
@devil-ira This works as expected. The real problem is `winit` on Nvidia GPU and Wayland can't get surface with Srgb colors, but this problem has been resolved in https://github.com/rust-windowing/glutin/pull/1435. After bevy migrates to the new `winit` the color should be correct. Bevy should work on Nvidia and Wayland with the new `winit`, even without this PR, but I think it's reasonable to get texture format at runtime other than pre-defined.

**Comment by VitalyAnkh** [comment]
Bevy also needs wgpu v0.14.0(https://github.com/gfx-rs/wgpu/releases/tag/v0.14.0) to have the `winit` Srgb issues fixed.

**Review by alice-i-cecile — APPROVED** [review]
Not a complete fix, but a necessary one.

bors r+
